libgalois/include/katana/Timer.h

```diff
@@ -104,6 +104,24 @@ class CondStatTimer<false> {
   uint64_t get_usec() const { return 0; }
 };
 
+template <typename T>
+class TimerGuard {
+public:
+  T timer;
+
+  template <typename... Args>
+  TimerGuard(Args&&... args) : timer(std::forward<Args>(args)...) {
```

Comment:
Shouldn't this just take a reference to a timer rather than construct a new timer?

Reply:
I didn't want to keep a reference in the field of this class because that can create stale reference issues, and would force the client code to use two lines, one for the timer and one for the guard. The timer classes are not copyable or movable, so there isn't a clean way I can see to do this other than what I did here.

Reply:
I was thinking that TimerGuard was similar in spirit to std::lock_guard: https://en.cppreference.com/w/cpp/thread/lock_guard/lock_guard so it would enable things like:

```cpp
void Func() {
  TimeAccumulator t;
  {
    TimerGuard guard(t);
    Fn1();
  }

  {
    TimerGuard guard(t);
    Fn2();
  }
}
```

but it sounds like you want

```cpp
void Func() {
  TimerGuard t(TimeAccumulator());
  Fn1();
}
```

?

Also would it make sense to add `[[nodiscard]]` to the constructor?

Reply:
Yes. I was implementing the latter. Because the majority of the time there is only one region, but honestly the extra line of code isn't too bad and your idea is more generic. My only concern is making sure that `TimerGuard t(TimeAccumulator());` is an error if we expect the former case, since it could cause stack corruption, maybe C++ handles this these days, I'll have to check.

I will reimplement with your approach and with `[[nodiscard]]`.